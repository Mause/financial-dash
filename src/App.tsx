import "./App.css";
import { definitions } from "./supabase";
import { useTable, useUser, useSignIn, useSignOut } from "react-supabase-fp";
import { pipe, constant } from "fp-ts/function";
import { toNullable } from "fp-ts/Option";
import * as RD from "@devexperts/remote-data-ts";
import useSWR from "swr";
import { useState } from "react";

function money(obj: { amount: number }) {
  return "$" + obj.amount / 100;
}

function App() {
  const result = useTable<
    definitions["Bill"] & {
      Vendor: definitions["Vendor"];
      Payment: Array<definitions["Payment"] & { Payer: definitions["Payer"] }>;
    }
  >(
    "Bill",
    `
    id,
    amount,
    Payment (
      id,
      amount,
      bankId,
      Payer (
        name
      )
    ),
    Vendor (
      id,
      name
    )`
  );
  console.log(result);
  const { data, error } = useSWR("https://launtel.vercel.app/api/transactions");

  const [signInResult, signIn] = useSignIn();
  const [, signOut] = useSignOut();
  const user = useUser();
  const [email, setEmail] = useState<string>();

  return (
    <div className="App">
      <header className="App-header">
        Financial Dash
        {pipe(
          signInResult,
          RD.fold(
            () => (
              <>
                <input
                  placeholder="Email"
                  required
                  onChange={(e) => setEmail(e.target.value)}
                />
                <button
                  onClick={(e) => {
                    e.preventDefault();
                    signIn({ email });
                  }}
                >
                  Log in
                </button>
              </>
            ),
            constant(<div>Signing in...</div>),
            (error) => <div>{error.toString()}</div>,
            () => (
              <button
                onClick={(e) => {
                  e.preventDefault();
                  signOut();
                }}
              >
                Sign out
              </button>
            )
          )
        )}
      </header>
      <p>
        {data?.length} transactions || {error?.toString()}
      </p>
      <p>{pipe(user, toNullable)?.email}</p>
      <p>
        {pipe(
          result,
          RD.fold3(
            constant(<div>Loading...</div>),
            (e) => <div>Query failed: {e}</div>,
            (result) => (
              <>
                <h1>Bills</h1>
                <div>
                  {result.map((row) => (
                    <div key={row.id}>
                      <h2>
                        #{row.id} — {money(row)} — {row.Vendor.name} (#
                        {row.Vendor.id})
                      </h2>
                      <ul>
                        {row.Payment.map((payment) => (
                          <li key={payment.id}>
                            {payment.Payer.name}
                            {" — "}
                            {money(payment)}
                            {" — "}
                            {payment.bankId ? "Paid" : "Unpaid"}
                          </li>
                        ))}
                      </ul>
                    </div>
                  ))}
                </div>
              </>
            )
          )
        )}
      </p>
    </div>
  );
}

export default App;
