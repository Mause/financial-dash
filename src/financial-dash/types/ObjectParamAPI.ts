import { ResponseContext, RequestContext, HttpFile } from '../http/http';
import * as models from '../models/all';
import { Configuration} from '../configuration'

import { InlineObject } from '../models/InlineObject';
import { InlineResponseDefault } from '../models/InlineResponseDefault';

import { ObservableDefaultApi } from "./ObservableAPI";
import { DefaultApiRequestFactory, DefaultApiResponseProcessor} from "../apis/DefaultApi";

export interface DefaultApiCreateInvoiceRequest {
    /**
     * 
     * @type InlineObject
     * @memberof DefaultApicreateInvoice
     */
    inlineObject: InlineObject
}

export class ObjectDefaultApi {
    private api: ObservableDefaultApi

    public constructor(configuration: Configuration, requestFactory?: DefaultApiRequestFactory, responseProcessor?: DefaultApiResponseProcessor) {
        this.api = new ObservableDefaultApi(configuration, requestFactory, responseProcessor);
    }

    /**
     * @param param the request object
     */
    public createInvoice(param: DefaultApiCreateInvoiceRequest, options?: Configuration): Promise<void> {
        return this.api.createInvoice(param.inlineObject,  options).toPromise();
    }

}
