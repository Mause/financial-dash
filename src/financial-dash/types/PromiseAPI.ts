import { ResponseContext, RequestContext, HttpFile } from "../http/http";
import * as models from "../models/all";
import { Configuration } from "../configuration";

import { InlineObject } from "../models/InlineObject";
import { InlineResponseDefault } from "../models/InlineResponseDefault";
import { ObservableDefaultApi } from "./ObservableAPI";

import {
  DefaultApiRequestFactory,
  DefaultApiResponseProcessor,
} from "../apis/DefaultApi";
export class PromiseDefaultApi {
  private api: ObservableDefaultApi;

  public constructor(
    configuration: Configuration,
    requestFactory?: DefaultApiRequestFactory,
    responseProcessor?: DefaultApiResponseProcessor
  ) {
    this.api = new ObservableDefaultApi(
      configuration,
      requestFactory,
      responseProcessor
    );
  }

  /**
   * @param inlineObject
   */
  public createInvoice(
    inlineObject: InlineObject,
    _options?: Configuration
  ): Promise<void> {
    const result = this.api.createInvoice(inlineObject, _options);
    return result.toPromise();
  }
}
