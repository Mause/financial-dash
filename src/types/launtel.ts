/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {
  "/api/transactions": {
    get: operations["getTransactions"];
  };
  "/api/usage": {
    get: operations["getUsage"];
  };
}

export interface components {
  schemas: {
    LauntelTransactionResponse: {
      perMonth: string;
    };
    UsageResponse: {
      usage: { [key: string]: unknown };
    };
  };
}

export interface operations {
  getTransactions: {
    responses: {
      /** Ok */
      default: {
        content: {
          "application/json": components["schemas"]["LauntelTransactionResponse"];
        };
      };
    };
  };
  getUsage: {
    responses: {
      /** Ok */
      default: {
        content: {
          "application/json": components["schemas"]["UsageResponse"];
        };
      };
    };
  };
}

export interface external {}
