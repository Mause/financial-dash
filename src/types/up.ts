/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {
  "/api/up": {
    get: operations["getUp"];
  };
}

export interface components {
  schemas: {
    UpAttributes: {
      description: string;
      message: string;
      createdAt: string | string;
      amount: components["schemas"]["Amount"];
    };
    UpTransaction: {
      id: string;
      attributes: components["schemas"]["UpAttributes"];
    };
    UpTransactionResponse: {
      items: components["schemas"]["UpTransaction"][];
    };
    Amount: {
      valueInBaseUnits: number;
      value: string;
    };
  };
  responses: {};
  parameters: {};
  requestBodies: {};
  headers: {};
}

export interface operations {
  getUp: {
    responses: {
      /** Ok */
      default: {
        content: {
          "application/json": components["schemas"]["UpTransactionResponse"];
        };
      };
    };
  };
}

export interface external {}
