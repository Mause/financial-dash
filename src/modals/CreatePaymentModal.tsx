import { useTable, useInsert } from "react-supabase-fp";
import * as RD from "@devexperts/remote-data-ts";
import { useState, ChangeEvent, FormEvent } from "react";
import { Modal, Button, Form, Notification } from "react-bulma-components";
import { SetB, Payment, Payer } from "../App";
import * as O from "fp-ts/Option";
import Axios from "axios";
import { constant } from "fp-ts/lib/function";
import { useToken } from "../auth";
import { InvoiceService } from "../invoice-ninja";

const invoiceService = new InvoiceService();

export function CreatePaymentModal(props: {
  setShow: SetB;
  bill: number;
  refresh: () => void;
}) {
  const [createPaymentResult, createPayment] = useInsert<Payment>("Payment");
  const [payers] = useTable<Payer>("Payer");
  const [payer, setPayer] = useState<number>();
  const [amount, setAmount] = useState<number>();
  const token = useToken();

  if (RD.isSuccess(createPaymentResult)) {
    props.setShow(false);
    props.refresh();
  }

  return (
    <Modal.Card
      renderAs="form"
      onSubmit={async (e: FormEvent<any>) => {
        e.preventDefault();

        const res = await invoiceService.createInvoice(
            RD.isSuccess(payers)
              ? payers.value.find((p) => p.id === payer)!.invoice_ninja_id
              : undefined,
            amount,
          {
            tokenProvider: {
              getToken() {
                return Promise.resolve(O.getOrElse(constant(""))(token));
              }
            },
          }
        );

        await createPayment({
          paidBy: payer,
          amount,
          paidFor: props.bill,
          invoice_ninja_id: res.id,
        });
      }}
    >
      <Modal.Card.Header>
        <Modal.Card.Title>Add Payment</Modal.Card.Title>
      </Modal.Card.Header>
      <Modal.Card.Body>
        {RD.isFailure(createPaymentResult) && (
          <Notification>{createPaymentResult.error}</Notification>
        )}
        <Form.Field>
          <Form.Control>
            <Form.Select
              loading={RD.isPending(payers)}
              onChange={(e: ChangeEvent<HTMLSelectElement>) => {
                setPayer(Number(e.target.value));
              }}
              required
            >
              <option value="">Select a payer</option>
              {RD.isSuccess(payers) &&
                payers.value.map((payer) => (
                  <option key={payer.id} value={payer.id}>
                    {payer.name}
                  </option>
                ))}
            </Form.Select>
          </Form.Control>
        </Form.Field>
        <Form.Field kind="addons">
          <Form.Control>
            <Button isStatic={true}>$</Button>
          </Form.Control>
          <Form.Control>
            <Form.Input
              name="amount"
              placeholder="Amount"
              required
              onChange={(e: ChangeEvent<HTMLInputElement>) => {
                setAmount(Number(e.target.value) * 100);
              }}
            ></Form.Input>
          </Form.Control>
        </Form.Field>
      </Modal.Card.Body>
      <Modal.Card.Footer>
        <Button type="submit">Create</Button>
      </Modal.Card.Footer>
    </Modal.Card>
  );
}
