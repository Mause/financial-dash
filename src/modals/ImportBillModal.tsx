import { useInsert } from "react-supabase-fp";
import * as RD from "@devexperts/remote-data-ts";
import useSWR from "swr";
import { useState, FormEvent } from "react";
import { Modal, Button, Form, Notification } from "react-bulma-components";
import { SetB, Bill } from "../App";

export function ImportBillModal(props: {
  setOpenImportBill: SetB;
  refresh: () => void;
}) {
  const [createBillResult, createBill] = useInsert<Bill>("Bill");
  const [month, setMonth] = useState<string>();
  const { data, error, isValidating } = useSWR<{
    perMonth: Record<string, { discounted: number }>;
  }>("https://launtel.vercel.app/api/transactions");

  if (RD.isSuccess(createBillResult)) {
    props.setOpenImportBill(false);
    props.refresh();
  }

  return (
    <Modal.Card
      renderAs="form"
      onSubmit={async (e: FormEvent<unknown>) => {
        e.preventDefault();

        if (!(month && data)) return;

        const amount = data.perMonth[month]?.discounted;
        if (amount == null) return;

        await createBill({
          vendor: 1,
          billDate: month + "-01",
          amount: Math.floor(amount * 100),
        });
      }}
    >
      <Modal.Card.Header>
        <Modal.Card.Title>Import Bill</Modal.Card.Title>
      </Modal.Card.Header>
      <Modal.Card.Body>
        {error && <Notification>{error}</Notification>}
        {RD.isFailure(createBillResult) && (
          <Notification>{createBillResult.error}</Notification>
        )}
        <Form.Field>
          <Form.Label>Select a transaction</Form.Label>
          <Form.Control>
            <Form.Select
              onChange={(e) => setMonth(e.target.value)}
              loading={isValidating}
              required
            >
              {Object.entries(data?.perMonth ?? {}).map(
                ([month, { discounted }]) => (
                  <option key={month} value={month}>
                    {month}
                    {" â€” $"}
                    {discounted}
                  </option>
                )
              )}
            </Form.Select>
          </Form.Control>
        </Form.Field>
      </Modal.Card.Body>
      <Modal.Card.Footer renderAs={Button.Group}>
        <Button type="submit">Import</Button>
      </Modal.Card.Footer>
    </Modal.Card>
  );
}
