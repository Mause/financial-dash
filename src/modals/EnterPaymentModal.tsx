import { useUpdate } from "react-supabase-fp";
import * as RD from "@devexperts/remote-data-ts";
import useSWRInfinite from "swr/infinite";
import { useState, FormEvent, MouseEvent } from "react";
import { Modal, Button, Form, Notification } from "react-bulma-components";
import { formatISO, parseISO } from "date-fns";
import { SetB, Payment, PaymentWithPayer } from "../App";
import { components } from "../types/up";
import { PaymentApi } from "../financial-dash";
import useApi from "../use_api";
import _ from "lodash";
import log from "../log";

function getKey(
  pageIndex: number,
  previousPageData: components["schemas"]["UpTransactionResponse"] | null
) {
  if (pageIndex === 0) {
    return "https://up.vc.mause.me/api/up?page[size]=100";
  }
  return previousPageData?.links?.next || null;
}

export function EnterPaymentModal(props: {
  setShowModal: SetB;
  selectedPayment: PaymentWithPayer;
  refresh: () => void;
}): JSX.Element {
  const [bankId, setBankId] = useState<string>();
  const [result, updatePayment] = useUpdate<Payment>("Payment");
  const { data, size, setSize, error, isValidating } =
    useSWRInfinite<components["schemas"]["UpTransactionResponse"]>(getKey);

  if (RD.isSuccess(result)) {
    props.setShowModal(false);
    props.refresh();
  }

  const paymentClient = useApi(PaymentApi);

  log.info({ data }, "Up transactions");

  return (
    <Modal.Card
      renderAs="form"
      onSubmit={async (e: FormEvent<unknown>) => {
        e.preventDefault();
        await paymentClient.postPayment({
          client_id: props.selectedPayment.Payer.invoice_ninja_id,
          amount: String(props.selectedPayment.amount!),
          invoice_id: props.selectedPayment.invoice_ninja_id,
          transaction_reference: bankId!,
        });
        await updatePayment({ bankId }, (query) =>
          query.eq("id", props.selectedPayment.id)
        );
      }}
    >
      <Modal.Card.Header>
        <Modal.Card.Title>
          Enter payment for {props.selectedPayment.Payer.name}
        </Modal.Card.Title>
      </Modal.Card.Header>
      <Modal.Card.Body>
        {error && <Notification>{JSON.stringify(error)}</Notification>}
        <Form.Label>Select a transaction</Form.Label>
        <Form.Field kind="addons">
          <Form.Control>
            <Button isStatic={true}>Page {size}</Button>
          </Form.Control>
          <Form.Control>
            <Form.Select
              required
              onChange={(e) => setBankId(e.target.value)}
              loading={isValidating}
            >
              <option value="">Select a transaction</option>
              {_.flatMap(data, (page) =>
                page.items.map((transaction) => (
                  <option key={transaction.id} value={transaction.id}>
                    {formatISO(parseISO(transaction.attributes.createdAt), {
                      representation: "date",
                    })}
                    {" — "}${transaction.attributes.amount.value}
                    {" — "}
                    {transaction.attributes.description}
                    {" — "}
                    {transaction.attributes.message}
                  </option>
                ))
              )}
            </Form.Select>
          </Form.Control>
          <Form.Control>
            <Button
              onClick={(e: MouseEvent) => {
                e.preventDefault();
                setSize(size + 1);
              }}
            >
              Load More
            </Button>
          </Form.Control>
        </Form.Field>
      </Modal.Card.Body>
      <Modal.Card.Footer renderAs={Button.Group}>
        <Button type="submit">Pay</Button>
      </Modal.Card.Footer>
    </Modal.Card>
  );
}
