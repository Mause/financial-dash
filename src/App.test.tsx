import "@testing-library/jest-dom";

import { act, render, RenderResult, screen } from "@testing-library/react";
import App, { BillCard, PaymentWithPayer } from "./App";
import { Provider } from "react-supabase-fp";
import { SupabaseClient } from "@supabase/supabase-js";
import { EnterPaymentModal } from "./modals/EnterPaymentModal";
import _ from "lodash";
import { SWRConfig } from "swr";
import { components } from "./types/up";

test("renders learn react link", async () => {
  await act(async () => {
    const expected = new ExpectPromises([[]]);
    const supa = {
      auth: {
        onAuthStateChange() {
          return {
            data: {
              unsubscribe: NOOP,
            },
          };
        },
        session() {
          return {};
        },
      },
      from() {
        return {
          select() {
            return expected;
          },
        };
      },
    };
    render(
      <Provider value={supa as unknown as SupabaseClient}>
        <App />
      </Provider>,
    );
    await expected.all();
  });

  const linkElement = screen.getByText(/loading/i);
  expect(linkElement).toBeInTheDocument();
});

class ExpectPromises {
  promises: Promise<unknown>[];
  resolvers: (() => void)[];

  constructor(responses: unknown[]) {
    this.promises = [];
    this.resolvers = [];
    for (const response of responses) {
      let resolve: ((v: unknown) => void) | undefined;
      const promise = new Promise<unknown>((_resolve) => {
        resolve = _resolve;
      });
      this.promises.push(promise);
      this.resolvers.push(resolve!.bind(promise, response));
    }
  }

  then() {
    const resolve = this.resolvers.pop();
    if (resolve) resolve();
  }

  async all(): Promise<void> {
    await Promise.all(this.promises);
  }
}

test("Bill", async () => {
  const expected = new ExpectPromises([2]);

  const supa = {
    from() {
      return {
        delete() {
          return { eq: () => expected, in: () => expected };
        },
      };
    },
  };
  let el: RenderResult;

  await act(async () => {
    el = render(
      <Provider value={supa as unknown as SupabaseClient}>
        <BillCard
          row={{
            id: 0,
            billDate: "2021-01-01",
            amount: 1111,
            Payment: [
              {
                id: 0,
                amount: 500,
                invoice_ninja_id: "aaa",
                Payer: { id: 0, name: "Fred", invoice_ninja_id: "" },
                paidFor: -1,
                paidBy: -1,
              },
            ],
            Vendor: { id: 0, name: "Synergy" },
          }}
          setSelectedPayment={NOOP}
          refresh={NOOP}
          setShowModal={NOOP}
        />
      </Provider>,
    );
  });
  expect(el!.container).toMatchSnapshot();

  await act(async () => {
    screen.getByText("Delete").click();
    await expected.all();
  });

  expect(el!.container).toMatchSnapshot();
});

test("EnterPaymentModal", async () => {
  const fetcher = async function (): Promise<
    components["schemas"]["UpTransactionResponse"]
  > {
    return {
      items: [
        {
          id: "0000",
          attributes: {
            amount: {
              value: "10.00",
              valueInBaseUnits: 1000,
            },
            description: "This is a description",
            createdAt: "2022-01-12T10:33",
            message: "This is a message",
          },
        },
      ],
      links: {},
    };
  };

  const payment: PaymentWithPayer = {
    id: 0,
    paidFor: 0,
    amount: 1500,
    paidBy: 0,
    bankId: "",
    invoice_ninja_id: "aaa",
    Payer: { id: 0, name: "Hello", invoice_ninja_id: "" },
  };
  let el: RenderResult<typeof import("@testing-library/dom/types/queries")>;
  await act(async () => {
    el = render(
      <SWRConfig value={{ fetcher }}>
        <EnterPaymentModal
          refresh={NOOP}
          setShowModal={NOOP}
          selectedPayment={payment}
        />
      </SWRConfig>,
    );
  });
  expect(el!.container).toMatchSnapshot();
});

const NOOP = _.identity;
