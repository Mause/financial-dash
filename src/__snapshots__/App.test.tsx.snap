// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bill 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <div
        class="card-header-title"
      >
        #
        0
         — 
        2021-01-01
         — 
        $11.11
         — 
        Synergy
         (#
        0
        )
        <button
          class="is-small button"
          tabindex="0"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      class="card-content"
    >
      <ul>
        <li>
          Fred
           — 
          $5
           — 
          <button
            class="is-small button"
            tabindex="0"
          >
            Unpaid
          </button>
        </li>
      </ul>
    </div>
    <div
      class="card-footer"
    />
  </div>
</div>
`;

exports[`Bill 2`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <div
        class="card-header-title"
      >
        #
        0
         — 
        2021-01-01
         — 
        $11.11
         — 
        Synergy
         (#
        0
        )
        <button
          class="is-small button"
          tabindex="0"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      class="card-content"
    >
      <ul>
        <li>
          Fred
           — 
          $5
           — 
          <button
            class="is-small button"
            tabindex="0"
          >
            Unpaid
          </button>
        </li>
      </ul>
    </div>
    <div
      class="card-footer"
    />
  </div>
</div>
`;
